### PR TITLE
Add Formula Column message, type, and test support

### PR DESCRIPTION
**Pull request by ian-legler**
Add new Formula Columns error messages.
Add new FormulaType implementations for Hyperlink and Invalid to allow IR translation of Dataverse types.
Update test files to clarify SQL support

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:43]: Do these types participate in analysis? 
Eg, If X is a HyperLink, does Eval("If(..., x, ...)") return a hyperlink type?

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:76]: >  throw new NotImplementedException($"Not implemented type: {type}"); [](http://example.com/codeflow?start=18&length=69)

Can we keep the exception?
Introducing Invalid here can cover up real issues.

**Comment by MikeStall** [comment]
> // Scalar + Table

Are these tests coming from somewhere else? or new tests?

---
Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TableStringfuncs.txt:1 in a989e48. [](commit_id = a989e485229dad464d0e80120b3a5b9fff9cc245, deletion_comment = False)

**Comment by ian-legler** [comment]
> > // Scalar + Table
> 
> Are these tests coming from somewhere else? or new tests?
> 
> Refers to: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/TableStringfuncs.txt:1 in [a989e48](https://github.com/microsoft/Power-Fx/commit/a989e485229dad464d0e80120b3a5b9fff9cc245). [](commit_id = [a989e48](https://github.com/microsoft/Power-Fx/commit/a989e485229dad464d0e80120b3a5b9fff9cc245), deletion_comment = False)

These new tests were added to string.txt.  I just moved them to their own file, similar to how math functions are done

File: src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj
Comment (lesaltzm) [inline src/tests/Microsoft.PowerFx.Core.Tests/Microsoft.PowerFx.Core.Tests.csproj:54]: Why are we excluding this *.txt file from the csproj?

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/HyperlinkType.cs
Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/HyperlinkType.cs:9]: It coerces to string, but I didn't think it's actually in use anywhere. Do we have more info on the scenarios for this type?

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/InvalidType.cs
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Public/Types/InvalidType.cs:18]: Should be implemented in ITypeVisitor.cs

Comment (ian-legler) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/InvalidType.cs:18]: Done

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs
Comment (ian-legler) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:43]: Yes.
These are being added to unblock IR Translation, which fails if it can't map the binding type to a FormulaType

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs
Comment (ian-legler) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:76]: I added a new function BuildOrInvalid that can return Invalid.  Existing calls to Build will still throw.

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs
Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/FormulaType.cs:43]: Where are they coming from? Does Dataverse register a hyperlink type?

File: src/libraries/Microsoft.PowerFx.Core/Public/Types/HyperlinkType.cs
Comment (ian-legler) [reply src/libraries/Microsoft.PowerFx.Core/Public/Types/HyperlinkType.cs:9]: The DType is used for URL formatted strings in Dataverse.  This is needed forIR translation to succeed, so the SQL compiler can reject them later.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/string.txt
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/string.txt:297]: Why is this a compiler error? The { } should be handled in the IR and the same as the existing concat operations.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/mathfuncs.txt
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/mathfuncs.txt:266]: Is this jsut a case of nyi?  We need to factor this file better for this. (in the future, not blocking this PR)

File: src/tests/Microsoft.PowerFx.Core.Tests/PublicSurfaceTests.cs
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Core.Tests/PublicSurfaceTests.cs:73]: > "Microsoft.PowerFx.Core.Public.Types.InvalidType", [](http://example.com/codeflow?start=15&length=51)

remove?

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/string.txt
Comment (ian-legler) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/string.txt:297]: String interpolation is behind a feature flag that isn't turned on in the compiler

Comment (MikeStall) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/string.txt:297]: answered offline: still under feature flag, and not lighting up in SQL yet.

**Review by MikeStall — DISMISSED** [review]
:shipit:

File: src/tests/Microsoft.PowerFx.Core.Tests/PublicSurfaceTests.cs
Comment (ian-legler) [reply src/tests/Microsoft.PowerFx.Core.Tests/PublicSurfaceTests.cs]: Done

**Review by MikeStall — APPROVED** [review]

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/mathfuncs.txt
Comment (ian-legler) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/mathfuncs.txt:266]: This is relatively new, and  not implemented in SQL yet

**Pull request merged**
